### ISSUE TITLE
Can we merge multiple glb using jgltf?

### ISSUE THREAD

**Issue by ritik619** [javagl/JglTF#83]
Hello 
I've two GLBs one containing animation and another containing model.
I have had success in applying animation to models in threeJs.
I want to replicate the same in my Kotlin App,

Will it be possible to do so via JglTF??

I've discovered two possibilities so far :
1. Creating a new GLB merging "model.glb" and "animation.glb" and then using it  ( I was able to do so using  "gltf-transform" lib but not available in kotlin/Java)
2. By Applying animation on the go to the model (Unable to do so till now using filament) 

I'm using google/filament  for viewing and applying animation currently
@javagl

**Comment by javagl** [comment]
That's not so easy to answer. Technically, the `GltfModelBuilder` _should_ offer the required functionality. But there is no pre-built function that can do this. The reason is, as usual, that there are some caveats when trying to merge multiple assets. 

Referring to your specific example, it sounds like ...

- one of the assets contains the `meshes/accessors/bufferViews/buffers` of the model itself (maybe also `materials/textures/images`)
- the other asset contains the `scenes/nodes/animations/accessors/bufferViews/buffers` of an animation

Then there is a crucial question. And this is the question that makes it "nearly impossible" to generically merge "arbitrary" assets. But it might be possible to answer it, for the given example, and making some assumptions:

How do you know which `node` contains the `mesh` of the model?

(All this becomes "nearly impossible" when the assets involve _skinning_, but I assume that this is not the case here...)

**Comment by javagl** [comment]
Not all operations might be totally trivial here. (I'm not deeply familiar with gltf-transform, but could imagine that it is more tailored for certain _transforming_ operations. There are some "obvious" transformation steps that are on the radar to be implemented with JglTF as well, but which ones _exactly_ is hard to say). If you have some code snippet that shows what you're currently doing in gltf-transform, then I might try to come up with something that is similar to that. Otherwise, I could try to create a few lines of code that show how certain operations _might_ be done with JglTF.

**Comment by ritik619** [comment]
`const io = new NodeIO();
    const animationDocument1: Document = await io.read(`src/animations/mixamoHipHopDancing.glb`);
    const avtarDocument: Document = await io.read(`src/character/Godi.glb`);
    const doc: Document = new Document();
    await doc.merge(animationDocument1).merge(avtarDocument);
    const skin = doc.getRoot().listSkins()[0];
    doc.getRoot().listAnimations().forEach(animation => {
      const animationChannels=animation.listChannels();
      animationChannels.forEach(channel => {
        //this curTargetNode is node of Animation
        const curTargetNode = channel.getTargetNode()?.getName();
        //Findin the final Target Node
        const newTargetNode = skin.listJoints().find(node => node.getName() === curTargetNode);
        if (newTargetNode) channel.setTargetNode(newTargetNode);
      });
    });

    await doc.transform(dedup(), resample(), prune());
    const buffer = doc.getRoot().listBuffers()[0];
    doc
      .getRoot()
      .listAccessors()
      .forEach(a => a.setBuffer(buffer));
    doc
      .getRoot()
      .listBuffers()
      .forEach((b, index) => (index > 0 ? b.dispose() : null));
    await io.write(`src/Gaiki/a.glb`, doc);`

**Comment by ritik619** [comment]
animationLink : [](https://drive.google.com/file/d/1ha-INnuwv7Ln9_aimnmuPs78hf9sNiHn/view?usp=share_link)
Model Link: [](https://drive.google.com/file/d/157wfp0CH7zZbHh5T4XdIA7ES_9i0KgT8/view?usp=share_link)
Output Model with Animation : [](https://drive.google.com/file/d/1WpxrPKqCK1aPfjFfT4RSA6SjZO2w2Pt6/view?usp=share_link)
@javagl

**Comment by javagl** [comment]
I noticed that there is probably a bug in the model building process. I have to analyze it a bit more, but a preliminary fix is in https://github.com/javagl/JglTF/pull/85

Based on the state in this PR, it should be possible to merge the GLBs that you linked to. With the given input files, the following code should create the desired `merged.glb`, with the animated character:

```Java
package de.javagl.jgltf.issues;

import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import de.javagl.jgltf.model.AnimationModel;
import de.javagl.jgltf.model.AnimationModel.Channel;
import de.javagl.jgltf.model.AnimationModel.Sampler;
import de.javagl.jgltf.model.GltfModel;
import de.javagl.jgltf.model.NodeModel;
import de.javagl.jgltf.model.SkinModel;
import de.javagl.jgltf.model.creation.GltfModelBuilder;
import de.javagl.jgltf.model.impl.DefaultAnimationModel;
import de.javagl.jgltf.model.impl.DefaultAnimationModel.DefaultChannel;
import de.javagl.jgltf.model.impl.DefaultGltfModel;
import de.javagl.jgltf.model.io.GltfModelReader;
import de.javagl.jgltf.model.io.GltfModelWriter;

public class JgltfIssue83_MergeGlbs
{
    public static void main(String[] args) throws Exception
    {
        String characterFileName = "data/mergeGlbFromIssue83/character.glb";
        String animationFileName = "data/mergeGlbFromIssue83/anim1.glb";
        String outputFileName = "data/mergeGlbFromIssue83/merged.glb";
        
        // Read the character- and animation model
        GltfModelReader gltfModelReader = new GltfModelReader();

        GltfModel gltfModelCharacter = gltfModelReader.read(
            Paths.get(characterFileName));
        System.out.println("Character model:");
        System.out.println(gltfModelCharacter);
        
        GltfModel gltfModelAnimation = gltfModelReader.read(
            Paths.get(animationFileName));
        System.out.println("Animation model");
        System.out.println(gltfModelAnimation);

        // Assume that there is exactly one skin model in the character:
        List<SkinModel> skinModels = gltfModelCharacter.getSkinModels();
        SkinModel skinModel = skinModels.get(0);

        // Convert the "old" animation models from the animations input
        // into "new" animation models that refer to the nodes from
        // the character input
        List<AnimationModel> oldAnimationModels =
            gltfModelAnimation.getAnimationModels();
        List<AnimationModel> newAnimationModels =
            new ArrayList<AnimationModel>();
        for (AnimationModel oldAnimationModel : oldAnimationModels)
        {
            List<Channel> newChannels = new ArrayList<Channel>();
            List<Channel> oldChannels = oldAnimationModel.getChannels();
            for (Channel oldChannel : oldChannels)
            {
                NodeModel channelNode = oldChannel.getNodeModel();
                String nodeName = channelNode.getName();

                // Try to find a node in the joints of the skin model
                // of the character input that has the same name as the
                // node of the animation channel of the animation input
                Optional<NodeModel> optionalNodeModel = skinModel.getJoints()
                    .stream()
                    .filter(n -> n.getName().equals(nodeName))
                    .findFirst();
                if (optionalNodeModel.isPresent())
                {
                    System.out.println("Adding animation for " + nodeName);
                    Sampler sampler = oldChannel.getSampler();
                    NodeModel nodeModel = optionalNodeModel.get();
                    String path = oldChannel.getPath();
                    DefaultChannel newChannel =
                        new DefaultChannel(sampler, nodeModel, path);
                    newChannels.add(newChannel);
                }
            }
            if (!newChannels.isEmpty())
            {
                DefaultAnimationModel newAnimationModel =
                    new DefaultAnimationModel();
                for (Channel newChannel : newChannels)
                {
                    newAnimationModel.addChannel(newChannel);
                }
                newAnimationModels.add(newAnimationModel);
            }
        }

        // Build the new model. First, add all elements from the
        // character model (EXCEPT for the "dummy" animations 
        // that it contained)
        GltfModelBuilder gltfModelBuilder = GltfModelBuilder.create();
        addAll(gltfModelBuilder, gltfModelCharacter);
        
        // Now, add the newly generated animation models
        gltfModelBuilder.addAnimationModels(newAnimationModels);

        // Build the whole thing and write the result
        DefaultGltfModel newModel = gltfModelBuilder.build();
        GltfModelWriter gltfModelWriter = new GltfModelWriter();
        gltfModelWriter.writeBinary(newModel,
            Paths.get(outputFileName).toFile());
    }

    private static void addAll(GltfModelBuilder b, GltfModel m)
    {
        b.addCameraModels(m.getCameraModels());
        b.addImageModels(m.getImageModels());
        b.addMaterialModels(m.getMaterialModels());
        b.addMeshModels(m.getMeshModels());
        b.addNodeModels(m.getNodeModels());
        b.addSceneModels(m.getSceneModels());
        b.addSkinModels(m.getSkinModels());
        b.addTextureModels(m.getTextureModels());
    }
}
```

I have to investigate the issue of the linked PR a bit more, and maybe add a (simple) test case. But maybe you can do some tests for your use case based on the state from the PR and the above code.

**Comment by ritik619** [comment]
Thank you very much @javagl 
After your previous comment i was able to figure out a way to add animation myself as well.
However I'm so thankful for your response.
Thank you for making a library for java gl
This is great!

**Comment by javagl** [comment]
Some bugs have been fixed as part of https://github.com/javagl/JglTF/pull/108 that is now released with version 2.0.4. A dedicated function to merge multiple glTFs does not exist yet, but further thoughts about this can be tracked in https://github.com/javagl/JglTF/issues/105
